autorest/models/name_converter.py

```diff
@@ -0,0 +1,186 @@
+        schema_python_name = schema['language']['python']['name']
+
+        schema_python_name = NameConverter._to_camel_case(schema_name)
+        schema['language']['python']['name'] = schema_python_name
+        for prop in schema.get('properties', []):
+            if self._seen_id(id(prop)):
+                continue
+            NameConverter._convert_language_default_name(schema=prop, pad_string='Property')
+
+    @staticmethod
+    def _convert_language_default_name(schema, **kwargs):
+        schema['language']['python'] = dict(schema['language']['default'])
+        schema_name = schema['language']['default']['name']
+        schema_python_name = schema['language']['python']['name']
+
+        schema_python_name = NameConverter._to_valid_python_name(schema_name, **kwargs)
+        schema['language']['python']['name'] = schema_python_name
+
+
+    @staticmethod
+    def _to_camel_case(name):
+        name_list = re.split('[^a-zA-Z\\d]', name)
+        name_list = [s[0].upper() + s[1:] if len(s) > 1 else s.upper()
+                            for s in name_list]
+        return ''.join(name_list)
+
+    @staticmethod
+    def _to_valid_python_name(name, **kwargs):
+        pad_string = kwargs.pop('pad_string', "")
+        if not name:
+            return pad_string
+        return NameConverter._to_python_case(
+            NameConverter._get_escaped_reserved_name(
+                NameConverter._to_valid_name(
+                    name.replace('-', '_'), '_'),
+                    pad_string
+                )
+            )
+
+    @staticmethod
+    def _to_python_case(name):
+        def replace_upper_characters(m):
+            match_str = m.group().lower()
+            if m.start() > 0 and name[m.start() - 1] == '_':
+                # we are good if a '_' already exists
+                return match_str
+            # the first letter should not have _
+            prefix = '_' if m.start() > 0 else ''
+
+            # we will add an extra _ if there are multiple upper case chars together
+            next_non_upper_case_char_location = m.start() + len(match_str)
+            if (len(match_str) > 2 and len(name) - next_non_upper_case_char_location > 1 and
+                name[next_non_upper_case_char_location].isalpha()):
+
+                return prefix + match_str[: len(match_str) - 1] + '_' + match_str[len(match_str) - 1]
+
+            return prefix + match_str
+
+        return re.sub("[A-Z]+", replace_upper_characters, name)
+
+    @staticmethod
+    def _get_escaped_reserved_name(name, append_value):
+        if name is None:
+            raise TypeError("The value for name can not be None")
+        if append_value is None:
+            raise TypeError("The value for append_value can not be None")
+        if name.lower() in reserved_words:
+            name += append_value
+        return name
+
+    @staticmethod
+    def _remove_invalid_characters(name, allowed_characters):
```

Comment:
What is this guy for? Please add unittests with all simple and corner case scenario and docstring.

Reply:
added